### ISSUE TITLE
[BUG]: Error executing process > 'NF_FLU:ILLUMINA:SUBTYPING_REPORT'

### ISSUE THREAD

**Issue by Codes1985** [CFIA-NCFAD/nf-flu#32]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Description of the Bug/Issue

Hello,

First of all, thank you very much for your efforts to continue to enhance the nf-flu workflow.  It is truly appreciated.

I have been testing each release against Illumina and Nanopore datasets for both influenza A and B.

I encountered an error during the subtyping report generation step (logs attached) runing nf-flu against an Illumina-based influenza B dataset.  I haven't been able to reproduce this error running release 3.3.0 on any other dataset.  I'm inclined to think it is related to parsing the specific metadata associated with the BLAST results for this sample.

I'll be sure to update this thread with any new information.

Thank you in advance.

EDIT: It definitely appears to be sample-specific.  Trying to isolate offending samples for further investigation.

### Nextflow command-line

```shell
nextflow nf-flu_v3.3.0/cfia-ncfad-nf-flu-3.3.0/workflow/main.nf --input IRVC20230711_SK_Illumina_fluB_validation_setup.csv --platform illumina --outdir IRVC20230711_SK_Illumina_fluB_validation_setup_nf-flu_results -profile singularity,slurm
```

### Error Message

```shell
[cd/928ea9] process > NF_FLU:ILLUMINA:SUBTYPING_R... [ 50%] 1 of 2, failed: 1...
[-        ] process > NF_FLU:ILLUMINA:SOFTWARE_VE... -
[d2/0af9b2] NOTE: Process `NF_FLU:ILLUMINA:SUBTYPING_REPORT` terminated with an error exit status (1) -- Execution is retried (1)
ERROR ~ Error executing process > 'NF_FLU:ILLUMINA:SUBTYPING_REPORT'

Caused by:
  Process `NF_FLU:ILLUMINA:SUBTYPING_REPORT` terminated with an error exit status (1)

Command executed:

  parse_influenza_blast_results.py \
   --flu-metadata 41415333-influenza.csv \
   --top 3 \
   --excel-report nf-flu-subtyping-report.xlsx \
   --pident-threshold 0.85 \
   sample36.blastn.txt sample46.blastn.txt sample38.blastn.txt sample28.blastn.txt sample30.blastn.txt sample44.blastn.txt sample26.blastn.txt sample6.blastn.txt sample32.blastn.txt sample40.blastn.txt sample14.blastn.txt sample10.blastn.txt sample22.blastn.txt sample18.blastn.txt sample24.blastn.txt sample42.blastn.txt sample16.blastn.txt sample34.blastn.txt sample8.blastn.txt sample20.blastn.txt sample4.blastn.txt sample2.blastn.txt sample12.blastn.txt
  ln -s .command.log parse_influenza_blast_results.log
  cat <<-END_VERSIONS > versions.yml
  "NF_FLU:ILLUMINA:SUBTYPING_REPORT":
     python: $(python --version | sed 's/Python //g')
  END_VERSIONS

Command exit status:
  1

Command output:
  (empty)

Command error:
  │ │                   ┆ null       ┆ null       ┆ Sequence   │               │ │
  │ │                   │           ┆            ┆        ┆        ┆   ┆       │ │
  │ │                   ┆            ┆            ┆ 7703 from  │               �� │
  │ │                   │           ┆            ┆        ┆        ┆   ┆       │ │
  │ │                   ┆            ┆            ┆ Patent     │               │ │
  │ │                   │           ┆            ┆        ┆        ┆   ┆       │ │
  │ │                   ┆            ┆            ┆ WO2007…    │               │ │
  │ │                   │ sample6_6 ┆ GN357980.1 ┆ 94.737 ┆ 57     ┆ … ┆ null  │ │
  │ │                   ┆ null       ┆ null       ┆ Sequence   │               │ │
  │ │                   │           ┆            ┆        ┆        ┆   ┆       │ │
  │ │                   ┆            ┆            ┆ 7744 from  │               │ │
  │ │                   │           ┆            ┆        ┆        ┆   ┆       │ │
  │ │                   ┆            ┆            ┆ Patent     │               │ │
  │ │                   │           ┆            ┆        ┆        ┆   ┆       │ │
  │ │                   ┆            ┆            ┆ WO2007…    │               │ │
  │ │                   │ sample6_6 ┆ GN357979.1 ┆ 93.333 ┆ 60     ┆ … ┆ null  │ │
  │ │                   ┆ null       ┆ null       ┆ Sequence   │               │ │
  │ │                   │           ┆            ┆        ┆        ┆   ┆       │ │
  │ │                   ┆            ┆            ┆ 7743 from  │               │ │
  │ │                   │           ┆            ┆        ┆        ┆   ┆       │ │
  │ │                   ┆            ┆            ┆ Patent     │               │ │
  │ │                   │           ┆            ┆        ┆        ┆   ┆       │ │
  │ │                   ┆            ┆            ┆ WO2007…    │               │ │
  │ │                   └───────────┴────────────┴────────┴────────┴───┴─────… │ │
  │ │  df_type_counts = shape: (0, 3)                                          │ │
  │ │                   ┌──────────┬────────┬────────┐                         │ │
  │ │                   │ Genotype ┆ counts ┆ N_type │                         │ │
  │ │                   │ ---      ┆ ---    ┆ ---    │                         │ │
  │ │                   │ str      ┆ u32    ┆ str    │                         │ │
  │ │                   ╞══════════╪════════╪════════╡                         │ │
  │ │                   └──────────┴────────┴────────┘                         │ │
  │ │          h_or_n = 'N'                                                    │ │
  │ │          is_iav = True                                                   │ │
  │ │ reg_h_or_n_type = '[Nn]'                                                 │ │
  │ │             seg = '6'                                                    │ │
  │ │     type_counts = shape: (3, 2)                                          │ │
  │ │                   ┌──────────┬────────┐                                  │ │
  │ │                   │ Genotype ┆ counts │                                  │ │
  │ │                   │ ---      ┆ ---    │                                  │ │
  │ │                   │ str      ┆ u32    │                                  │ │
  │ │                   ╞══════════╪════════╡                                  │ │
  │ │                   │ B        ┆ 86     │                                  │ │
  │ │                   │ Victoria ┆ 64     │                                  │ │
  │ │                   │ Yamagata ┆ 33     │                                  │ │
  │ │                   └──────────┴────────┘                                  │ │
  │ │       type_name = 'N_type'                                               │ │
  │ │   type_to_count = []                                                     │ │
  │ ╰──────────────────────────────────────────────────────────────────────────╯ │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  IndexError: list index out of range

Work dir:
  validation_results/nf-flu_v3.3.0/Illumina_FluB/IRVC20230711_SK_Illumina_fluB_validation_setup_nf-flu_results/work/cd/928ea9d2849072e0835391fa054a0c

Tip: view the complete command output by changing to the process work dir and entering the command `cat .command.out`

 -- Check '.nextflow.log' file for details
```

### Workflow Version

v3.3.0; revision: 91a5d05f86

### Nextflow Executor

Slurm

### Nextflow Version

23.04.1

### Java Version

openjdk version "20-internal" 2023-03-21
OpenJDK Runtime Environment (build 20-internal-adhoc..src)
OpenJDK 64-Bit Server VM (build 20-internal-adhoc..src, mixed mode, sharing)

### Hardware

HPC Cluster

### Conda/Container Engine

Singularity

### Additional context

[slurm-26930422_mod.log](https://github.com/CFIA-NCFAD/nf-flu/files/12072941/slurm-26930422_mod.log)
[command.log](https://github.com/CFIA-NCFAD/nf-flu/files/12072943/command.log)
[nextflow_mod.log](https://github.com/CFIA-NCFAD/nf-flu/files/12072947/nextflow_mod.log)

**Comment by nhhaidee** [comment]
`Genotype` column from updated NCBI Influenza db has strange values such as `Yamagata`, `Victoria` (please see below for more details), that's why `type_to_count` is empty, will check with @peterk87 when he back.

`\\((H\\d+N\\d+|H7N2|N9|H6N7|(H6N4)|European|H14N2|H16N9|N3|H12N8|H11N2|H15N5|H8N3|H1N6|H15N7|H9N7|H12N4|H1N7|n1|H5N6|h4n2|H1N9|H13N4|H11N9/N2|A|H3Nx|unidentified|H2N4|H11N8|H11N9|HxN6|Victoria|HxN2|H4|H3N9|HxN3|h5n1|H11N7|H15N4|H10N5|H13N2|H1Nx|H2N6|H6N6|H14N4|N8|unknown|H7Nx|H13N?|H4N6|H12N1|H1N2v|H5N9|H14N3|H1N(untyped)|H8N4|H9N8|H11N1|H4N5|H11|Yamagata-like|H12N9|H11N|H16|H10N7|H12|mixed|H5N7|H9N?|H2N?|H11N3|H8N1|H5N1|N6|H5N4|H9N1,9|H3N6|HxN1|H6N1|H13N3|H4N7|H14|H3N5|H3N2v|H1N4|H7N8|H6N3|H9N4|H3N(untyped)|H15N6|H5|B|H5N3|H10|H10N2|H4N1|H7N3|H8N7|H9N2|H6N1,4|H9Nx|H9|H3N1|U5575-1|H2N1|H16N3|H2N9|H10N4|Mixed|N1|Influenza B|N7|H3|HIN2|H8N2|H6N9|H8Nx|H11N5|H8N8|H7N7|H4N3|h5n8|H3N2|N5|HPR0|N4|H1|H5N2?|H7|H?N?|H14N7|H6Nx|H13|H12N3|H2N5|H7N6|H3H2|H8N5|H6N8|HxNx|H12N6|H1N2|H10N9|H3N8|H2N3|H5N8|H9N9|H13N9|H0N1|H4N8|H19|H10N3|H3N7|H4N2|H1N5|H?N6|H7N4|Yamagata|H14N6|H11N4|H1n2|H14N8|H10N6|H12N5|H3N3|H11N?|H1N8|H8N6|H14N5|H1N1pdm09|H4Nx|H15N9|H15N2|H6N5|H2N7|H5N?|H91|H17N10|H?N4|H15N8|H6N?|H12N2|H4N4|H9N3|H6N4|European genotype|H?N1|H5N5|H4N4,8|H11N6|H3N4|H?N2|H13N8|H2N8|H18N11|H2|H5n8|H7N2NSB|H9N6|H7N5|N4N6|H4N9|H8|H1N3|North American|H3N-|H13N1|H13N6|H10N1|(H1N1)|N2|H9N1,4|H16N8|H7N?|H9N5|H1N?|H2N2|H6N|H7N1|H6N2|H6|H12N7|H10N8|H1N1|H5N2|H5Nx|H3N?|H01N2|H7N9|H9N1)\\)`

**Comment by peterk87** [comment]
@nhhaidee the values are correct, but the parsing of them needs to be adjusted for non IAV support, that is, wrap the regex matching and H/N parsing in a try catch and handle more gracefully than just an error causing the script to exit. I don't would recommend just showing the values that come up for IBV. This should be an easy fix.

**Comment by nhhaidee** [comment]
@peterk87 OK. The sample is IBV but line of code below returns `iav = True`, will fix it

https://github.com/CFIA-NCFAD/nf-flu/blob/b0d6575b6d0715e78ecc093297ec20cd08960747/bin/parse_influenza_blast_results.py#L248

**Comment by peterk87** [comment]
There may be genus or species info retained in the new metadata table. Some columns of info were dropped since they didn't have much meaningful info.

**Comment by nhhaidee** [comment]
The `GenBank_Title` column has species info, we check this column as well. I'll check this column carefully

        if df_top_seg_matches.select(pl.col("Genotype").is_null().all())[0, 0] and df_top_seg_matches.select(
                pl.col("GenBank_Title").str.contains('Influenza A').all())[0,0]:
            is_iav = True
        else:
            is_iav = False

**Comment by Codes1985** [comment]
Thanks a lot of looking into this, guys!  On the topic of metadata, I also ran into an edge case where the HA subtype was "missed" for a particular sample because the top BLAST result did not actually have the HA subtype listed in its corresponding metadata.  Something to consider.

**Comment by ayooluwaB** [comment]
Hi, 
Thanks for all your help in 
this issue persists for fluB on Nanopore as well. I updated/did a new pull before running the pipeline. See below for the error gotten: 

[6e/f909c0] process > NF_FLU:NANOPORE:SUBTYPING_R... [100%] 2 of 2, failed: 2...
[-        ] process > NF_FLU:NANOPORE:SOFTWARE_VE... -
[-        ] process > NF_FLU:NANOPORE:MULTIQC        -
Oops... Pipeline execution stopped with the following message: │ │                   ┆ null       ┆ null       ┆ Sequence   │               │ │
│ │                   │            ┆            ┆        ┆        ┆   ┆      │ │
│ │                   ┆            ┆            ┆ 7715 from  │               �� │
│ │                   │            ┆            ┆        ┆        ┆   ┆      │ │
│ │                   ┆            ┆            ┆ Patent     │               │ │
│ │                   │            ┆            ┆        ┆        ┆   ┆      │ │
│ │                   ┆            ┆            ┆ WO2007…    │               │ │
│ │                   │ 23700909_6 ┆ GN357960.1 ┆ 93.333 ┆ 60     ┆ … ┆ null │ │
│ │                   ┆ null       ┆ null       ┆ Sequence   │               │ │
│ │                   │            ┆            ┆        ┆        ┆   ┆      │ │
│ │                   ┆            ┆            ┆ 7724 from  │               │ │
│ │                   │            ┆            ┆        ┆        ┆   ┆      │ │
│ │                   ┆            ┆            ┆ Patent     │               │ │
│ │                   │            ┆            ┆        ┆        ┆   ┆      │ │
│ │                   ┆            ┆            ┆ WO2007…    │               │ │
│ │                   │ 23700909_6 ┆ GN357959.1 ┆ 93.333 ┆ 60     ┆ … ┆ null │ │
│ │                   ┆ null       ┆ null       ┆ Sequence   │               │ │
│ │                   │            ┆            ┆        ┆        ┆   ┆      │ │
│ │                   ┆            ┆            ┆ 7723 from  │               │ │
│ │                   │            ┆            ┆        ┆        ┆   ┆      │ │
│ │                   ┆            ┆            ┆ Patent     │               │ │
│ │                   │            ┆            ┆        ┆        ┆   ┆      │ │
│ │                   ┆            ┆            ┆ WO2007…    │               │ │
│ │                   └────────────┴────────────┴────────┴────────┴───┴────… │ │
│ │  df_type_counts = shape: (0, 3)                                          │ │
│ │                   ┌──────────┬────────┬────────┐                         │ │
│ │                   │ Genotype ┆ counts ┆ N_type │                         │ │
│ │                   │ ---      ┆ ---    ┆ ---    │                         │ │
│ │                   │ str      ┆ u32    ┆ str    │                         │ │
│ │                   ╞══════════╪════════╪════════╡                         │ │
│ │                   └──────────┴────────┴────────┘                         │ │
│ │          h_or_n = 'N'                                                    │ │
│ │          is_iav = True                                                   │ │
│ │ reg_h_or_n_type = '[Nn]'                                                 │ │
│ │             seg = '6'                                                    │ │
│ │     type_counts = shape: (3, 2)                                          │ │
│ │                   ┌──────────┬────────┐                                  │ │
│ │                   │ Genotype ┆ counts │                                  │ │
│ │                   │ ---      ┆ ---    │                                  │ │
│ │                   │ str      ┆ u32    │                                  │ │
│ │                   ╞══════════╪════════╡                                  │ │
│ │                   │ B        ┆ 86     │                                  │ │
│ │                   │ Victoria ┆ 64     │                                  │ │
│ │                   │ Yamagata ┆ 33     │                                  │ │
│ │                   └──────────┴────────┘                                  │ │
│ │       type_name = 'N_type'                                               │ │
│ │   type_to_count = []                                                     │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
╰──────────────────────────────────────────────────────────────────────────────╯
IndexError: list index out of range
[a8/e4d103] NOTE: Process `NF_FLU:NANOPORE:SUBTYPING_REPORT_BCF_CONSENSUS` terminated with an error exit status (1) -- Execution is retried (1)
ERROR ~ Error executing process > 'NF_FLU:NANOPORE:SUBTYPING_REPORT_BCF_CONSENSUS'

Caused by:
  Process `NF_FLU:NANOPORE:SUBTYPING_REPORT_BCF_CONSENSUS` terminated with an error exit status (1)

Command executed:

  parse_influenza_blast_results.py \
   --flu-metadata 41415333-influenza.csv \
   --top 3 \
   --excel-report nf-flu-subtyping-report.xlsx \
   --pident-threshold 0.85 \
   19912214.blastn.txt 19912004.blastn.txt 19907392.blastn.txt 18915205.blastn.txt 19913443.blastn.txt 19909832.blastn.txt 19907930.blastn.txt 23723927.blastn.txt 19912580.blastn.txt 19910303.blastn.txt 199132                            23.blastn.txt 19905018.blastn.txt 19911747.blastn.txt 19925564.blastn.txt 23700909.blastn.txt 23714751.blastn.txt 19913908.blastn.txt 19922663.blastn.txt 18915038.blastn.txt 19908450.blastn.txt 19923723.blastn                            .txt 19913444.blastn.txt 19920391.blastn.txt
  ln -s .command.log parse_influenza_blast_results.log
  cat <<-END_VERSIONS > versions.yml
  "NF_FLU:NANOPORE:SUBTYPING_REPORT_BCF_CONSENSUS":
     python: $(python --version | sed 's/Python //g')
  END_VERSIONS



Thanks for your help in resolving this.

**Comment by peterk87** [comment]
Thanks @Codes1985 and @ayooluwaB for reporting your issue. 

We have just released [3.3.2](https://github.com/CFIA-NCFAD/nf-flu/releases/tag/3.3.2) where this issue with IBV genotype/subtype parsing has been fixed (#33). Test IBV sequence data for Illumina and Nanopore has been added so that should help with quickly catching any regressions in the future.

Thanks for your patience and we hope the new release works better for you! If not, please feel free to reopen the issue and provide as much info as you can so we can resolve the issue once and for all.
